checker/src/call_graph.rs

```diff
@@ -720,13 +684,175 @@ impl CallGraph {
+        let type_regex = Regex::new(r"(?P<type>[^<>,\s]+)").unwrap();
+        for caps in type_regex.captures_iter(&s2) {
+            let cap_type = &caps["type"];
+            if !COLLECTION_TYPES.contains(&cap_type) {
+                base_types.push(format!("[{}]", cap_type).into_boxed_str());
+            }
+        }
+        base_types
+    }
+
+    /// Simplify a type according to these heuristics:
+    /// - &t == t
+    /// - mut t == t
+    /// - [t] > t
+    fn simplify_single_type(&self, name: &str) -> SimpleType {
+        let mut base_type = name.to_owned();
+        if base_type.contains("mut ") {
+            base_type = base_type.replace("mut ", "");
+        }
+        if base_type.contains('&') {
+            base_type = base_type.replace("&", "");
+        }
+        if base_type.contains('[') && base_type.contains(']') {
+            base_type = base_type.replace("[", "").replace("]", "");
+            SimpleType::Collection(base_type.into_boxed_str())
+        } else {
+            SimpleType::Base(base_type.into_boxed_str())
+        }
+    }
+
+    /// Simplify a type that may be a collection
+    fn simplify_type(&self, name: &str) -> Vec<SimpleType> {
+        let is_collection_type: bool = COLLECTION_TYPES.iter().any(|x| name.contains(x));
+        if is_collection_type {
+            let base_types = self.extract_iterator_type(name);
+            base_types
+                .iter()
+                .map(|x| self.simplify_single_type(x))
+                .collect::<Vec<SimpleType>>()
+        } else {
+            vec![self.simplify_single_type(name)]
+        }
+    }
+
+    /// Check whether two types are nominally equivalent or
+    /// equivalent according to input type relations
+    fn simple_type_eq(&self, t1: &str, t2: &str, eq_map: &HashMap<Box<str>, Box<str>>) -> bool {
+        if t1 == t2 {
+            return true;
+        } else {
+            if let Some(map_t1) = eq_map.get(t2) {
+                if t1 == map_t1.as_ref() {
+                    return true;
+                }
+            }
+            if let Some(map_t2) = eq_map.get(t1) {
+                if t2 == map_t2.as_ref() {
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
+    /// Derive all type relations
+    fn derive_all_relations(
+        &self,
+        type_map: &HashMap<TypeId, Box<str>>,
+        eq_map: &HashMap<Box<str>, Box<str>>,
+    ) -> HashSet<TypeRelation> {
+        let mut simple_types = Vec::<(Box<str>, SimpleType)>::new();
+        for (_, t) in type_map.iter() {
+            let simple_ts = self.simplify_type(t.as_ref());
+            for simple_t in simple_ts.iter() {
+                simple_types.push((t.to_owned(), simple_t.to_owned()));
+            }
+        }
+        let mut relations = HashSet::<TypeRelation>::new();
+        // Initialize the set of pairs with the existing equality relations
+        let mut pairs = eq_map
```

Comment:
I don't see the point of pairs.

Reply:
The pairs set is used to avoid redundant EqType(a,b) relations by keeping track of the pairs (a,b) and (b,a) that have been added so far. It is initialized with the existing EqType relations that come from user input.

Reply:
Normalization and insert seem to do that already.

Reply:
Unfortunately no, because the set of relations is initially empty; it doesn't contain the input type relations. However, I will refactor this so that the optimization you've suggested does work.